xocto/fields/postgres/ranges.py

```diff
@@ -0,0 +1,220 @@
+        Transform an exclusive upper bound to an inclusive bound.
+
+        DateRanges (as all discrete ranges) are always stored as [) in the database, even if the
+        input was []. [X, Y] is equivalent to [X, Y+1) in the database. When we read the value
+        from the database, we need to subtract 1 day from the upper bound to return it to inclusive.
+        """
+        if value.upper_inc:
+            return value.upper
+        return value.upper - datetime.timedelta(days=1)
+
+
+class FiniteDateTimeRangeField(pg_fields.DateTimeRangeField):
+    """
+    A DateTimeRangeField with Inclusive-Exclusive [) bounds that aren't infinite.
+
+    Accepts and returns xocto.ranges.FiniteDatetimeRange objects. Values are always
+    timezone-aware, and will be converted to the timezone specified in django settings.
+    """
+
+    def get_prep_value(
+        self, value: Optional[ranges.FiniteDatetimeRange]
+    ) -> Optional[pg_ranges.DateTimeTZRange]:
+        if value is None:
+            return None
+        if not isinstance(value, ranges.FiniteDatetimeRange):
+            raise TypeError(
+                "FiniteDateTimeRangeField may only accept FiniteDatetimeRange objects."
+            )
+        return pg_ranges.DateTimeTZRange(
+            lower=value.start, upper=value.end, bounds="[)"
+        )
+
+    def from_db_value(
+        self,
+        value: Optional[pg_ranges.DateTimeTZRange],
+        expression: object,
+        connection: object,
+    ) -> Optional[ranges.FiniteDatetimeRange]:
+        if value is None:
+            return None
+        return ranges.FiniteDatetimeRange(
+            start=localtime.as_localtime(value.lower),
+            end=localtime.as_localtime(value.upper),
+        )
+
+    def to_python(self, value: Optional[str]) -> Optional[ranges.FiniteDatetimeRange]:
+        if value is None:
+            return None
+        obj = json.loads(value)
+        return ranges.FiniteDatetimeRange(
+            start=localtime.as_localtime(self.base_field.to_python(obj["start"])),
+            end=localtime.as_localtime(self.base_field.to_python(obj["end"])),
+        )
+
+    def value_to_string(self, obj: models.Model) -> Optional[str]:
+        value: Optional[ranges.FiniteDatetimeRange] = self.value_from_object(obj)
+        if value is None:
+            return None
+        if not isinstance(value, ranges.FiniteDatetimeRange):
+            raise TypeError(
+                "FiniteDateTimeRangeField may only accept FiniteDatetimeRange objects."
+            )
+        base_field = self.base_field
+        start = pg_utils.AttributeSetter(base_field.attname, value.start)
+        end = pg_utils.AttributeSetter(base_field.attname, value.end)
+        return json.dumps(
+            {
+                "start": base_field.value_to_string(start),
+                "end": base_field.value_to_string(end),
+            }
+        )
+
+
+class HalfFiniteDateTimeRangeField(pg_fields.DateTimeRangeField):
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Why no `HalfFiniteDateRangeField`

You the the `FiniteDateRange` to represent an inclusive-inclusive `Date` range with a mandatory start and end. Why not also have an inclusive-exclusive `Date` range with a potentially open end?

Reply:
Good question, but already addressed in the PR description:

> Notably missing is: HalfFiniteDateRangeField because we don't have a corresponding type in xocto.ranges. We already have usage of a half-open daterange field within our main application, so it would be a useful addition later.